tsl/src/data_node.c

```diff
@@ -1416,6 +1429,279 @@ data_node_detach(PG_FUNCTION_ARGS)
 	PG_RETURN_INT32(removed);
 }
 
+enum Anum_show_conn
+{
+	Anum_alter_data_node_node_name = 1,
+	Anum_alter_data_node_host,
+	Anum_alter_data_node_port,
+	Anum_alter_data_node_database,
+	Anum_alter_data_node_options,
+	_Anum_alter_data_node_max,
+};
+
+#define Natts_alter_data_node (_Anum_alter_data_node_max - 1)
+
+static HeapTuple
+create_alter_data_node_tuple(TupleDesc tupdesc, const char *node_name, List *options)
+{
+	Datum values[Natts_alter_data_node];
+	bool nulls[Natts_alter_data_node] = { false };
+	ListCell *lc;
+	ArrayBuildState *arrbuilder = initArrayResult(TEXTOID, CurrentMemoryContext, false);
+
+	MemSet(nulls, false, sizeof(nulls));
+
+	values[AttrNumberGetAttrOffset(Anum_alter_data_node_node_name)] = CStringGetDatum(node_name);
+
+	foreach (lc, options)
+	{
+		DefElem *elem = lfirst(lc);
+
+		if (strcmp("host", elem->defname) == 0)
+		{
+			values[AttrNumberGetAttrOffset(Anum_alter_data_node_host)] =
+				CStringGetTextDatum(defGetString(elem));
+		}
+		else if (strcmp("port", elem->defname) == 0)
+		{
+			int port = atoi(defGetString(elem));
+			values[AttrNumberGetAttrOffset(Anum_alter_data_node_port)] = Int32GetDatum(port);
+		}
+		else if (strcmp("dbname", elem->defname) == 0)
+		{
+			values[AttrNumberGetAttrOffset(Anum_alter_data_node_database)] =
+				CStringGetDatum(defGetString(elem));
+		}
+		else
+		{
+			StringInfo opt = makeStringInfo();
+			appendStringInfo(opt, "%s=%s", elem->defname, defGetString(elem));
+			accumArrayResult(arrbuilder,
+							 CStringGetTextDatum(opt->data),
+							 false,
+							 TEXTOID,
+							 CurrentMemoryContext);
+		}
+	}
+
+	values[AttrNumberGetAttrOffset(Anum_alter_data_node_options)] =
+		makeArrayResult(arrbuilder, CurrentMemoryContext);
+
+	return heap_form_tuple(tupdesc, values, nulls);
+}
+
+/*
+ * Switch data node to use for queries on chunks.
+ *
+ * When reset=false it will switch from the given data node to another one,
+ * but only if the data node is currently used for queries on the chunk.
+ *
+ * When reset=true it will switch to the given data node, if it is "primary"
+ * for the chunk (according to the current partitioning configuration).
+ */
+static void
+switch_data_node_on_chunks(const ForeignServer *datanode, bool reset)
```

Comment:
I am finding it a bit difficult to follow the `reset` semantics here, why not to call it `available`everywhere starting from the intitial call in the `data_node_alter()`?

Reply:
I guess I figured the parameter is semantically about what the function should do and not the status of the data node.

Reply:
Fixed.